[flang][OpenMP] Fix firstprivate not working with lastprivate in DO SIMD

KrxGu opened the pull request:
This fixes a bug where firstprivate was ignored when the same variable had both firstprivate and lastprivate clauses in a do simd construct.

What was broken:
```
integer :: a
a = 10
!$omp do simd firstprivate(a) lastprivate(a)
do i = 1, 1
   print *, a  ! Should print 10, but printed garbage/0
   a = 20
end do
!$omp end do simd
print *, a  ! Correctly prints 20
```

Inside the loop, [a] wasn't being initialized from the firstprivate clause—it just had whatever uninitialized value was there.

The fix:

In genCompositeDoSimd(), we were using simdItemDSP to handle privatization for the whole loop nest. This only looked at SIMD clauses and missed the firstprivate from the DO part. Changed it to use wsloopItemDSP instead, which handles both DO clauses (firstprivate, lastprivate) correctly.

One line change in OpenMP.cpp

Tests added:

Lowering test to check MLIR generation
Runtime test to verify the actual values are correct
<img width="740" height="440" alt="image" src="https://github.com/user-attachments/assets/fa911ea8-2024-4edf-b710-52c10659742e" />


Fixes #168306

Commit:
[flang][OpenMP] Fix firstprivate in do simd composite constructs

Remove duplicate privatization in do simd composites. Only the wsloop
should create private allocations; simd reuses them per OpenMP spec.

Fixes #168306

tblah commented on flang/test/Lower/OpenMP/do-simd-firstprivate-lastprivate.f90:
```diff
@@ -0,0 +1,87 @@
+  ! CHECK:      omp.wsloop
+  ! CHECK-SAME: private(@{{.*}}firstprivate{{.*}} %{{.*}} -> %[[FIRSTPRIV_A:.*]] : !fir.ref<i32>)
+  ! CHECK-NEXT: omp.simd
+  ! CHECK-SAME: private(@{{.*}} %{{.*}} -> %[[PRIV_A:.*]], @{{.*}} %{{.*}} -> %[[PRIV_I:.*]] : !fir.ref<i32>, !fir.ref<i32>)
```
Please could you match the neames of the private variables so that it is easier to keep track of which is used where.

tblah reviewed (commented):
I'm not sure about your interpretation of the spec. There is no firstprivate clause for the simd construct so the firstprivate cannot apply to the simd. Running flang without your patch it looks like we correctly apply firstprivate and lastprivate to the wsloop but only lastprivate to simd.

This results in LLVM codegen where there are three private variables: the firstprivate copy of a for `omp do` - this is unused and would be removed by the optimizer; the private copy of `a` used in the body of `omp simd`; and the private copy of the loop iteration variable. As you observed, this means that the value of `a` inside the loop is uninitialized memory.

If this is a bug then I think the bug is that there are two completely separate declarations and memory for `a`. This is still visible in your lit test.

Commit:
[flang][OpenMP] Update tests for corrected do simd privatization

Update test expectations after fixing do simd composite construct
privatization. Wsloop now correctly handles all privatization clauses;
simd no longer creates duplicate private allocations.

tblah reviewed (approved):
Thanks for the update. LG

Commit:
[flang][OpenMP] Add CHECK-NOT to ensure simd has no private clause

Add explicit negative checks to verify that omp.simd does not have
privatization clauses, ensuring we catch any future regressions.
